### PR TITLE
Create new github action to handle auto creation and destruction of test VMs

### PR DESCRIPTION
**Pull request by NdibeRaymond**
see #1002

### CODE REVIEW

**Comment by NdibeRaymond** [comment]
update on this @tuxology , this took slightly longer than the expected 1 week time because of unexpected issues and trying to joggle it with work. Currently having timeout and memory issues with `react-script build`. Once I figure out a way to solve that we should be good

File: test_vm_deployment/deploy_test_fullstack.sh [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+echo "removing old frontend folder"
+rm -rf /home/zubhub_frontend
+echo "removing old frontend folder"
+
+echo "remove old backend folder"
+rm -rf /home/zubhub_backend
+echo "removing old backend folder"
+
+echo "copying new frontend folder"
+cp -r /home/zubhub/zubhub_frontend/ /home/zubhub_frontend/
```
Comment (yokwejuste) [inline test_vm_deployment/deploy_test_fullstack.sh:26]: Maybe we need to get a copy of the frontend `env` file too?
```suggestion
cp -r /home/zubhub/zubhub_frontend/ /home/zubhub_frontend/
cp -r /home/zubhub/zubhub_frontend/zubhub/.env.example /home/zubhub_frontend/zubhub/.env
```

File: .github/workflows/create_destroy_test_vm.yaml [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+##################################################################################
+      - name: Install doctl
+        uses: digitalocean/action-doctl@v2
+        with:
+          token: ${{ secrets.DO_ACCESS_TOKEN }}
+
+      - name: Create new droplet and DNS records
+        run: |
+
+          NEW_DROPLET_NAME=zubhub-test-${{ github.event.pull_request.number }}
```
Comment (yokwejuste) [inline .github/workflows/create_destroy_test_vm.yaml:38]:
```suggestion
          NEW_DROPLET_NAME=${{ env.DROPLET_NAME_PREFIX }}${{ github.event.pull_request.number }}
```

File: .github/workflows/create_destroy_test_vm.yaml [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+        ) && contains(github.event.pull_request.labels.*.name, 'create-test-vm')
+      ) ||
+      (
+        github.event.action == 'labeled' &&
+        github.event.label.name == 'create-test-vm'
+      )
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout files
+        uses: actions/checkout@v2
```
Comment (yokwejuste) [inline .github/workflows/create_destroy_test_vm.yaml:24]: Upgrade actions/checkout from v2 to v4 for faster checkout and better performance. V4 utilizes the native Git client, reducing resource usage and improving speed by up to 30%.

```suggestion
        uses: actions/checkout@v4
```

File: .github/workflows/create_destroy_test_vm.yaml [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+          EOF
+
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v1
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+        
+      - name: Login to DockerHub
+        uses: docker/login-action@v1
```
Comment (yokwejuste) [inline .github/workflows/create_destroy_test_vm.yaml:113]: Upgrade docker/login-action from v1 to v3! V3 boasts improved support for more registries, smoother multi-registry workflows, and a simpler, more consistent API. Ditch v1 for a streamlined and powerful login experience.

```suggestion
        uses: docker/login-action@v3
```

File: .github/workflows/create_destroy_test_vm.yaml [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+
+      - name: Build and push frontend
+        id: docker_build_frontend
+        uses: docker/build-push-action@v2
```
Comment (yokwejuste) [inline .github/workflows/create_destroy_test_vm.yaml:120]: The version 5 has similar implementation and more optimized in terms of resources usage.

```suggestion
        uses: docker/build-push-action@v5
```

File: .github/workflows/create_destroy_test_vm.yaml [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+          REACT_APP_DEV_VIDEO_FOLDER_NAME=dev_videos
+          REACT_APP_VIDEO_UPLOAD_PRESET_NAME=video_upload_preset
+          REACT_APP_DEV_VIDEO_UPLOAD_PRESET_NAME=dev_video_upload_preset
+          EOF
+
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v1
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
```
Comment (yokwejuste) [inline .github/workflows/create_destroy_test_vm.yaml:110]: Boost your builds with docker/setup-buildx-action v3! Enjoy smoother multi-platform builds, leverage BuildKit's full potential, and simplify your workflow with an updated API.
```suggestion
        uses: docker/setup-buildx-action@v3
```

File: .github/workflows/create_destroy_test_vm.yaml [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+          REACT_APP_DOSPACE_ACCESS_SECRET_KEY=
+          REACT_APP_VIDEO_UPLOAD_URL=
+          REACT_APP_VIDEO_FOLDER_NAME=videos
+          REACT_APP_DEV_VIDEO_FOLDER_NAME=dev_videos
+          REACT_APP_VIDEO_UPLOAD_PRESET_NAME=video_upload_preset
+          REACT_APP_DEV_VIDEO_UPLOAD_PRESET_NAME=dev_video_upload_preset
+          EOF
+
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v1
```
Comment (yokwejuste) [inline .github/workflows/create_destroy_test_vm.yaml:107]: Here we have some little updates for more virtualization support.

```suggestion
        uses: docker/setup-qemu-action@v3
```

File: .github/workflows/create_destroy_test_vm.yaml [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+          context: ./zubhub_backend/
+          file: ./zubhub_backend/compose/celery/Dockerfile
+          push: true
+          tags: unstructuredstudio/zubhub-test_celery:latest
+      - name: Image digest
+        run: echo ${{ steps.docker_build.outputs.digest }}
+
+      - name: Build and push media service
+        id: docker_build_media
+        uses: docker/build-push-action@v2
```
Comment (yokwejuste) [inline .github/workflows/create_destroy_test_vm.yaml:153]:
```suggestion
        uses: docker/build-push-action@v5
```

File: .github/workflows/create_destroy_test_vm.yaml [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+          context: ./zubhub_backend/
+          file: ./zubhub_backend/compose/web/prod/Dockerfile
+          push: true
+          tags: unstructuredstudio/zubhub-test_web:latest
+      - name: Image digest
+        run: echo ${{ steps.docker_build.outputs.digest }}
+
+      - name: Build and push celery worker
+        id: docker_build_celery
+        uses: docker/build-push-action@v2
```
Comment (yokwejuste) [inline .github/workflows/create_destroy_test_vm.yaml:142]:
```suggestion
        uses: docker/build-push-action@v5
```

File: .github/workflows/create_destroy_test_vm.yaml [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+          context: ./zubhub_frontend/zubhub/
+          file: ./zubhub_frontend/zubhub/Dockerfile.prod
+          push: true
+          tags: unstructuredstudio/zubhub-test_frontend:latest
+      - name: Image digest
+        run: echo ${{ steps.docker_build.outputs.digest }}
+
+      - name: Build and push django api
+        id: docker_build_web
+        uses: docker/build-push-action@v2
```
Comment (yokwejuste) [inline .github/workflows/create_destroy_test_vm.yaml:131]:
```suggestion
        uses: docker/build-push-action@v5
```

**Review by yokwejuste — COMMENTED** [review]
### Summary

#### Improvements implemented:
- DRY Principle: Leveraged variables to eliminate redundancy and improve code clarity.
- Action Upgrades: Updated actions to their latest versions for improved performance and security.
- Enhanced Variable Usage: Introduced additional variables to simplify configuration and enhance maintainability.

**Pull request closed without merging**